packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -705,10 +705,46 @@ - (int64_t)onPlayerSetup:(FVPVideoPlayer *)player frameUpdater:(FVPFrameUpdater
   return textureId;
 }
 
+// this same function is also in camera_avfoundation
```

Comment:
This is the least important information in the comment, so should be last as a final note.

Reply:
And it's not actually true, because they are slightly different due to `clearOptions`. So this should just be removed.

Reply:
Before it was the same (it was removed due to a comment about it not being used in camera). It was supposed to be in a single place but I could not find a way to share source files between plugins so I put it twice for the time being until it is possibly moved elsewhere into a single place. I think such a comment makes sense because without this context making certain changes to it in one plugin without changing it also in other plugins can result in it no longer working as intended (like simple change in camera to call it on capture session queue instead of main, that change may seem reasonable without context). Maybe I should change it a little like "..similar function..." and that these two need to be "synchronized" resp. on the same queue.

Reply:
The comments and implementation strategy seem overly focused on these two plugins to me. There could be 50 plugins in the ecosystem that set global audio state, only two of which we happen to control, so any solution that relies on these two plugins being exactly synchronized doesn't scale.

I would like the solution (and its comments) here to focus on things that are generally true:
- `AVAudioSession` is not explicitly documented as being thread safe, so we should only use it from the main thread. We can mention that other plugin or application code could be using `AVAudioSession.sharedInstance` as well, and that we have to hope that they are doing the same.
- We should upgrade rather than stomp the state because other code may be setting the global state as well, and we don't want to remove options they may have set.

Neither of those statements require exact synchronization with other code, and both generalize to the entire ecosystem. (Although as mentioned earlier, the real ecosystem-friendly solution is to move away from doing this in individual plugins at all, and leaving it for apps to do.)

Reply:
It is not enough to just use `AVAudioSession` on the main thread, something like this would not work well:

``` dart
dispatch_sync(main_queue, () => old_options = AVAudioSession.sharedInstance.categoryOptions);
new_options = old_options | AVAudioSessionCategoryOptionMixWithOthers;
dispatch_sync(main_queue, () => AVAudioSession.sharedInstance.categoryOptions = new_options);
```

Reply:
> AVAudioSession is not explicitly documented as being thread safe

Btw `category` and `categoryOptions` are atomic properties and I doubt that capture session is changing them on the main thread when `automaticallyConfiguresApplicationAudioSession` is true.

Reply:
Just for curiosity, here it is changed by capture session and it is not possible that it happened on the main thread:

``` objc
      dispatch_async(dispatch_get_main_queue(), ^{
        NSLog(@"%@", AVAudioSession.sharedInstance.category);
        dispatch_sync(_captureSessionQueue, ^{
          [_audioCaptureSession addOutput:_audioOutput];
        });
        NSLog(@"%@", AVAudioSession.sharedInstance.category);
      });
```

```
AVAudioSessionCategoryPlayback
AVAudioSessionCategoryPlayAndRecord
```

Reply:
Is this the last change request still not fulfilled after the latest commits? Should that comment be further rewritten to not explicitly mention the other plugin but something more general?